### DISCUSSION TITLE
Serial port COM3 for file transfer (using file dialog/TCP bridge).

### DISCUSSION THREAD

**Discussion by cr1901** [dbalsom/marty#49] [Ideas]
Normally, when I have a PC with only a serial port and install media, I do a bootstrap from [tcom.com](http://www.columbia.edu/kermit/ftp/a/tcomtxt.asm) or a [BASIC Kermit](http://www.columbia.edu/kermit/ftp/old/basic/kermit.bas) to upload additional files to my machine.

COM2 connects to a real port (which is great!), so you can emulate a file transfer with a null modem cable connected to _another_ serial port on your local machine. However, I'd like a simpler setup.

Would a COM3 port that is attached to a listening TCP port that sends/receives raw bytes be something you'd be interested in? I am willing to prototype this feature myself.

**Comment by dbalsom** [comment]
the idea occurred to me, and I had an idea for a TCP/IP modem using some kind of hayes commands to 'dial' IP addresses

Just noting you can already accomplish this sort of thing with the right os support, see
https://com0com.sourceforge.net/ and its com2tcp module, you can even do com->ssh with stunnel. https://www.stunnel.org/

and i believe on linux it's even easier as you can redirect a serial port with netcat.

**Reply by cr1901** [reply]
Indeed, I am on Windows. I almost mentioned com0com, but remembered that last time I used it there weren't any signed drivers (a _while_ ago). This seems to have changed, and I'm running it fine on unmodified 64-bit Windows 10.

> and i believe on linux it's even easier as you can redirect a serial port with netcat.

Do you mean [`socat`](https://linux.die.net/man/1/socat)? I don't think `netcat` can create ~~serial ports~~ device nodes, but `socat` can. It's been a _while_ since I've done that. If I have a *nix system I can test this on, I will try and reply w/ results.

In any case, there's solutions for the two major classes of OS that `martypc` supports (*nix and Windows). So regardless of what you decide to do wrt TCP/IP and serial (I'm curious about your modem idea- custom hardware :o?), I'm not stuck.

**Reply by dbalsom** [reply]
no custom hardware, just another emulated device.  similar to how the mouse is implemented - the mouse just converts mouse events from the window manager into mouse protocol bytes and sends them to through emulated COM1 serial port.

the only real advantage of having the virtual modem is being able to send it commands from inside the emulator; and reliving the experience of dialing a BBS.  Consider an internal 'phone book' that maps phone numbers to IP:port.  I dunno, the idea just sounded fun.

You'd want to have raw serial TCP connections as well, for applications that don't give you an interactive terminal.

**Comment by dbalsom** [comment]
In any case, the serial implementation in MartyPC is sort of barebones at the moment; i implemented just enough to get procomm plus to connect to a serial shell on a raspberry pi.  If you do any file transfers i'd be curious to see how well it performs.   Checkit diagnostics fails several serial port tests so I have some work to do there.  The UART i'm emulating is also pretty primitive.

**Reply by cr1901** [reply]
I don't have any particularly _comprehensive_ feedback to give at the moment, but I will say that I don't find `com0com` particularly reliable for any binaries > 1kB. I'm noticing several failure modes:

1. "phantom" null characters sneak in to the front of the binary when I do `COPY COM2 FOO.TXT`
2. It takes several `^Z` from my terminal program on the other end (Tera Term) before MS-DOS honors EOF, for `COPY COM2 FOO.TXT`.<sup>1</sup>
3. `TCOM.COM` sometimes truncates files > 10kB. However, `TCOM.COM` doesn't actually have an activity indicator, so for all I know I just got incredibly unlikely and ended the xfer near the end of the file.

Eventual goal is to get KERMIT on there and demonstrate a full bootstrap from nothing but a DOS diskette. I _have_ successfully transferred 1 program after keying in `TCOM.COM` (~180 chars) manually.

>In any case, the serial implementation in MartyPC is sort of barebones at the moment

Is there any way to debug the serial port I/O registers, and perhaps view the characters received?

1. Okay, `COPY` is in fact completely unreliable for doing even ASCII xfers if you want the file unchanged. BY default, MS-DOS serial port is "cooked" mode when doing a `COPY`, and _it'll rewrite CRLF as LF_!! However, I do find the serial port behavior weird thanks to 1 and 2. So I'd be happy to take a look while I look for my notes on the awful CRLF conversion.

**Reply by dbalsom** [reply]
>Is there any way to debug the serial port I/O registers, and perhaps view the characters received?

not at the moment, but I could add another debug window for serial.

**Reply by cr1901** [reply]
>so for all I know I just got incredibly unlikely and ended the xfer near the end of the file.

Just tried this again, and a > 10kB was truncated in much the same way. If you add a debug window for serial, and I can see what martypc "see"s, I can get a better idea of what's going on. Some of what I'm seeing (LF being stripped) is [DOS's fault](https://github.com/microsoft/MS-DOS/blob/master/v2.0/source/STRIN.ASM#L38-L39), but unsure of the rest.

**Reply by dbalsom** [reply]
how about a serial logfile that just dumps every byte that goes in or out?

**Reply by cr1901** [reply]
Works for me. Is it possible to add an option to log I/O port writes to the serial port as well (just so I know the status of the port)?

**Reply by dbalsom** [reply]
sure.

**Reply by dbalsom** [reply]
@cr1901 
Sorry it's been so long, but you can see from the 0.2.0 changelog i've been very busy.
0.2.1 does have a basic IO monitor

![image](https://github.com/dbalsom/martypc/assets/7229541/ca1eb0ae-d032-45ef-a60d-eee86a225319)

so you can at least see port reads/writes.  I still need to add serial logging and redirection to files and named pipes.  What has been holding me up on serial logging, is I really want a general iostream service i can inject into devices that can then request a logging stream or data stream, and the service will handle the type of stream as appropriate.  Nothing too crazy, just requires some thought. I think it will be a good addition for 0.2.1

**Reply by cr1901** [reply]
@dbalsom No worries, I'm patient :P. A logging service coupled w/ I/O port status should give me enough of a picture to figure out where truncation is happening.

Note that I've been busy myself (wrote a bunch of code in a new programming language, and the language changes quickly enough that it takes appreciable bandwidth to maintain my code). But I still have my `martypc` checkout. I will need to add checking out serial xfers to my TODO list.

If I haven't by the 0.2.1 release, feel free to yell :P.

**Reply by dbalsom** [reply]
Still pondering the proper general approach to device logging, but I just completed a major overhaul of MartyPC's serial emulation, so I wonder if things are any better.

There's also a first pass at a serial debug window.

![serial_stats](https://github.com/dbalsom/martypc/assets/7229541/ff97b41c-7e3a-4457-b30a-0e304b7944ef)

you can watch number go up.

**Comment by cr1901** [comment]
I've gotten the `v0.2.3` branch to compile and run, but there was a perf regression for me in dc4b4f7 according to bisect. Apparently the `wgpu` backend chosen changed (from NVIDIA Quadro to DX12) and performance (FPS) more than halved.

I acknowledge my hardware is older (Thinkpad W520), but seeing as it worked fine before this commit, I want to investigate further.

Also, what paint program is that?

**Reply by dbalsom** [reply]
Apparently it's not just integrated video adapters that got the opengl treatement. See the workaround in release notes. Basically you want to make an environment variable:

WGPU_BACKEND=gl

Eventually I might add in auto-fallback for this. I don't know why wgpu doesn't do it itself...

**Reply by dbalsom** [reply]
oh and that's PC Paintbrush.

**Reply by cr1901** [reply]
Okay, `WGPU_BACKEND=gl` fixed that. Indeed, if I had read the release notes, I would've noticed the fix. But who actually reads the manual, CHANGELOG, or release notes before diving right in? :)

I'll test to see if I can do a full multi-kb xfer over the serial line this weekend- on vacation today.

**Comment by cr1901** [comment]
I acknowledge there is a lot going on in this screenshot, but I wanted to get as much on-screen as possible for future-me. It looks like at least `CTTY` is working with the following options in `com0com` and Tera Term:

![image](https://github.com/dbalsom/martypc/assets/6418027/9b011cb1-40fe-4b56-a3fc-ed05b27b8781)

Of all these, I think only the following two are _absolutely required_ for `CTTY` to work:

* "Use Ports Class" must be checked in `com0com` setup. Ports of other types (`CNCA0`, `CNCB0`) don't seem to be enumerated.
* Flow control must be disabled in your terminal application. My [own experience](https://www.wdj-consulting.com/blog/ethflop/#fn-3) is that DOS doesn't assert `RTS`, so the terminal app will just buffer chars to oblivion waiting for `CTS` to go low. (However, DOS _does_ look at its `CTS` before deciding to send chars).

**Comment by cr1901** [comment]
Okay, so I've tried a `TCOM.COM` transfer with [this file](https://github.com/user-attachments/files/16042238/ckboo.txt). According to Windows, this file is 3,392 bytes:

![image](https://github.com/dbalsom/martypc/assets/6418027/33ef5bee-2594-47f8-a12e-39fb1f70fa20)

I have provided `TCOM.COM` on this disk image (no copyright stuff on it, so it should be fine): [scratch.zip](https://github.com/user-attachments/files/16042247/scratch.zip). It should be drive B, and DOS 2.1 should be drive A (probably can't legally distribute that, sorry :P).

`TCOM.COM` is designed to read from `COM2` at 300 baud, and write to either stdout or a file if redirected. I have the same setup for `com0com` as in my previous comment. Tera Term setup is also identical except 300 baud. I have Tera Term (`COM23`) and `martypc`'s `COM2` (`COM22` host) connected via `com0com`.

_From reboot_, if I do `TCOM > FOO.TXT` and then do a "File > Send File > ckboo.txt" from Tera Term's menu, I will receive a file that's 3,241bytes long, an RX Count of 3,322, and 85 RX<sup>1</sup> Overruns:
![image](https://github.com/dbalsom/martypc/assets/6418027/aae2346d-ddd1-4423-87df-cdd2ed872261)

I'm not sure where 3,241 comes from, but I have a good idea where the 3,322 vs 3,392 is coming from; according to `wc`, `ckboo.txt` is 70 lines. CRLF must be being rewritten to CR somewhere. Although I've [mentioned prior](https://github.com/dbalsom/martypc/discussions/49#discussioncomment-6421205) that DOS will do CRLF rewriting, `TCOM.COM` [uses](http://www.columbia.edu/kermit/ftp/a/tcomtxt.asm) `INT 0x14`, which I don't think would be subject to CRLF rewriting.

_I can deterministically reproduce these results across runs of `TCOM.COM` with the same file, even without rebooting. According to DOS' `COMP`, Identical files are created, so the characters being stripped/overrun are the same between runs._

## Footnotes
1. I think RX Overrun will unconditionally mark the first received after reboot as an overrun, because it's "technically correct" that the PC didn't read the port before receiving that character :). However, for my purposes I'm ignoring it. Additionally, it looks like "RX Last Byte" is only updated on overrun... or at the very least becomes `0x00` if an overrun didn't happen for the current byte. Is this correct/intended?

**Reply by dbalsom** [reply]
Can you retrieve the received file so we can compare?

**Reply by cr1901** [reply]
Somehow, the `scratch.dsk` I loaded is missing the FAT header, but DOS happily finds files on it anyway, with some errors according to `CHKDSK`- TIL!

I fixed this by formatting/copying to a new image, and then extracted the received file by mounting on a remote Linux box: [foo.txt](https://github.com/user-attachments/files/16046387/foo.txt).

Here is a hexdump of foo.txt. You can safely ignore the `^C\r\n` at the end; I believe that to be an artifact of Tera Term.

```
00000000: 424f 4f20 4649 4c45 2046 4f52 4d41 542c  BOO FILE FORMAT,
00000010: 2045 4e43 4f44 494e 4720 414e 4420 4445   ENCODING AND DE
00000020: 434f 4449 4e47 0d0d 4d61 6e79 204b 6572  CODING..Many Ker
00000030: 6d69 7420 7072 6f67 7261 6d73 2061 7265  mit programs are
00000040: 206f 6674 656e 2064 6973 7472 6962 7574   often distribut
00000050: 6564 2075 7369 6e67 2061 2073 7065 6369  ed using a speci
00000060: 616c 2065 6e63 6f64 696e 6720 6361 6c6c  al encoding call
00000070: 6564 0d22 626f 6f22 2028 7368 6f72 7420  ed."boo" (short 
00000080: 666f 7220 2262 6f6f 7473 7472 6170 2229  for "bootstrap")
00000090: 2066 6f72 6d61 742c 2064 6576 656c 6f70   format, develop
000000a0: 6564 2065 7370 6563 6961 6c6c 7920 666f  ed especially fo
000000b0: 7220 6469 7374 7269 6275 7469 6f6e 206f  r distribution o
000000c0: 660d 4b65 726d 6974 2062 696e 6172 6965  f.Kermit binarie
000000d0: 7320 6f76 6572 206e 6574 776f 726b 732c  s over networks,
000000e0: 2065 2d6d 6169 6c2c 2061 6e64 2063 6f6d   e-mail, and com
000000f0: 6d75 6e69 6361 7469 6f6e 206c 696e 6573  munication lines
00000100: 2e0d 0d41 2063 6f6d 6d6f 6e20 7072 6163  ...A common prac
00000110: 7469 6365 2069 7320 746f 2065 6e63 6f64  tice is to encod
00000120: 6520 2e45 5845 2061 6e64 206f 7468 6572  e .EXE and other
00000130: 2062 696e 6172 7920 6669 6c65 7320 696e   binary files in
00000140: 746f 2070 7269 6e74 6162 6c65 0d63 6861  to printable.cha
00000150: 7261 6374 6572 732c 2073 7563 6820 6173  racters, such as
00000160: 2068 6578 6164 6563 696d 616c 2064 6967   hexadecimal dig
00000170: 6974 732c 2066 6f72 2074 7261 6e73 706f  its, for transpo
00000180: 7274 6162 696c 6974 792e 2020 4120 7369  rtability.  A si
00000190: 6d70 6c65 2022 6865 7822 0d65 6e63 6f64  mple "hex".encod
000001a0: 696e 6720 7265 7375 6c74 7320 696e 2074  ing results in t
000001b0: 776f 2063 6861 7261 6374 6572 7320 7065  wo characters pe
000001c0: 7220 382d 6269 7420 6269 6e61 7279 2062  r 8-bit binary b
000001d0: 7974 652c 2070 6c75 7320 4352 4c46 7320  yte, plus CRLFs 
000001e0: 6164 6465 640d 6576 6572 7920 3830 2028  added.every 80 (
000001f0: 6f72 206c 6573 7329 2068 6578 2063 6861  or less) hex cha
00000200: 7220 7468 6520 6669 6c65 2074 6f20 7061  r the file to pa
00000210: 7373 2074 6872 6f75 6768 0d63 6172 642d  ss through.card-
00000220: 6f72 6965 6e74 6564 206e 6574 776f 726b  oriented network
00000230: 7320 6c69 6b65 2042 4954 4e45 542e 2020  s like BITNET.  
00000240: 4120 6865 7820 6669 6c65 2069 7320 7468  A hex file is th
00000250: 6572 6566 6f72 6520 6d6f 7265 2074 6861  erefore more tha
00000260: 6e20 7477 6963 650d 6173 206c 6172 6765  n twice.as large
00000270: 2061 7320 7468 6520 6f72 6967 696e 616c   as the original
00000280: 2062 696e 6172 7920 6669 6c65 2e0d 0d41   binary file...A
00000290: 202e 424f 4f20 6669 6c65 2069 7320 6120   .BOO file is a 
000002a0: 6d6f 7265 2063 6f6d 7061 6374 2c20 6275  more compact, bu
000002b0: 7420 736f 6d65 7768 6174 206d 6f72 6520  t somewhat more 
000002c0: 636f 6d70 6c69 6361 7465 642c 2065 6e63  complicated, enc
000002d0: 6f64 696e 672e 2020 4576 6572 790d 7468  oding.  Every.th
000002e0: 7265 6520 6269 6e61 7279 2062 7974 6573  ree binary bytes
000002f0: 2028 3234 2062 6974 7329 2061 7265 2073   (24 bits) are s
00000300: 706c 6974 2075 7020 696e 746f 2066 6f75  plit up into fou
00000310: 7220 362d 6269 7420 6279 7465 7320 7769  r 6-bit bytes wi
00000320: 7468 2034 3820 2841 5343 4949 0d63 6861  th 48 (ASCII.cha
00000330: 7261 6374 6572 2022 3022 2920 6164 6465  racter "0") adde
00000340: 6420 746f 2065 6163 682c 2072 6573 756c  d to each, resul
00000350: 7469 6e67 2069 6e20 666f 7572 2041 5343  ting in four ASC
00000360: 4949 2063 6861 7261 6374 6572 7320 7261  II characters ra
00000370: 6e67 696e 6720 6672 6f6d 0d22 3022 2028  nging from."0" (
00000380: 4153 4349 4920 3438 2920 746f 2022 6f22  ASCII 48) to "o"
00000390: 2028 4153 4349 4920 3131 3129 2c20 7769   (ASCII 111), wi
000003a0: 7468 2043 524c 4673 2061 6464 6564 2061  th CRLFs added a
000003b0: 7420 6f72 206e 6561 7220 2263 6f6c 756d  t or near "colum
000003c0: 6e20 3736 222e 0d54 6865 2072 6573 756c  n 76"..The resul
000003d0: 7469 6e67 2066 696c 6520 7369 7a65 2077  ting file size w
000003e0: 6f75 6c64 2074 6865 7265 666f 7265 2062  ould therefore b
000003f0: 6520 6162 6f75 7420 342f 3320 7468 6520  e about 4/3 the 
00000400: 2e20 2054 6869 730d 6973 2073 7469 6c6c  .  This.is still
00000410: 2071 7569 7465 206c 6172 6765 2c20 736f   quite large, so
00000420: 202e 424f 4f20 6669 6c65 7320 616c 736f   .BOO files also
00000430: 2063 6f6d 7072 6573 7320 636f 6e73 6563   compress consec
00000440: 7574 6976 6520 6e75 6c6c 2028 7a65 726f  utive null (zero
00000450: 290d 6279 7465 732e 2020 5570 2074 6f20  ).bytes.  Up to 
00000460: 3738 2063 6f6e 7365 6375 7469 7665 206e  78 consecutive n
00000470: 756c 6c73 2061 7265 2063 6f6d 7072 6573  ulls are compres
00000480: 7365 6420 696e 746f 2074 776f 2063 6861  sed into two cha
00000490: 7261 6374 6572 732e 2020 5469 6c64 650d  racters.  Tilde.
000004a0: 2822 7e22 2920 6973 2074 6865 206e 756c  ("~") is the nul
000004b0: 6c2d 636f 6d70 7265 7373 696f 6e20 6c65  l-compression le
000004c0: 6164 2d69 6e2c 2061 6e64 2074 6865 2066  ad-in, and the f
000004d0: 6f6c 6c6f 7769 6e67 2063 6861 7261 6374  ollowing charact
000004e0: 6572 2069 6e64 6963 6174 6573 0d68 6f77  er indicates.how
000004f0: 206d 616e 7920 6e75 6c6c 7320 6172 6520   many nulls are 
00000500: 7265 7072 6573 656e 7465 6420 2873 7562  represented (sub
00000510: 7472 6163 7420 3438 2066 726f 6d20 7468  tract 48 from th
00000520: 6973 2063 6861 7261 6374 6572 2773 2041  is character's A
00000530: 5343 4949 0d76 616c 7565 292e 2020 466f  SCII.value).  Fo
00000540: 7220 696e 7374 616e 6365 2022 7e41 2220  r instance "~A" 
00000550: 6d65 616e 7320 3137 2063 6f6e 7365 6375  means 17 consecu
00000560: 7469 7665 206e 756c 6c73 3b20 227e 7e22  tive nulls; "~~"
00000570: 206d 6561 6e73 2037 3820 6f66 2074 6865   means 78 of the
00000580: 6d2e 0d52 6570 6561 7465 6420 6e75 6c6c  m..Repeated null
00000590: 7320 6172 6520 7665 7279 2063 6f6d 6d6f  s are very commo
000005a0: 6e20 696e 2062 696e 6172 7920 6578 6563  n in binary exec
000005b0: 7574 6162 6c65 2066 696c 6573 2e0d 0d34  utable files...4
000005c0: 2d66 6f72 2d33 2065 6e63 6f64 696e 6720  -for-3 encoding 
000005d0: 636f 6d62 696e 6564 2077 6974 6820 6e75  combined with nu
000005e0: 6c6c 2063 6f6d 7072 6573 7369 6f6e 2072  ll compression r
000005f0: 6564 7563 6573 2074 6865 2073 697a 6520  educes the size 
00000600: 6f66 696c 6520 746f 2061 7070 726f 7869  ofile to approxi
00000610: 6d61 7465 6c79 2074 6865 2073 616d 6520  mately the same 
00000620: 7369 7a65 2061 7320 7468 6520 6f72 6967  size as the orig
00000630: 696e 616c 202e 4558 4520 6669 6c65 2c20  inal .EXE file, 
00000640: 616e 640d 736f 6d65 7469 6d65 7320 6576  and.sometimes ev
00000650: 656e 2073 6d61 6c6c 6572 2e20 2054 6865  en smaller.  The
00000660: 2066 6972 7374 206c 696e 6520 6f66 2061   first line of a
00000670: 202e 424f 4f20 6669 6c65 2069 7320 7468   .BOO file is th
00000680: 6520 6e61 6d65 2028 696e 2070 6c61 696e  e name (in plain
00000690: 0d74 6578 7429 206f 6620 7468 6520 6f72  .text) of the or
000006a0: 6967 696e 616c 2066 696c 652e 2020 4865  iginal file.  He
000006b0: 7265 2773 2077 6861 7420 7468 6520 6669  re's what the fi
000006c0: 7273 7420 6665 7720 6c69 6e65 7320 6f66  rst few lines of
000006d0: 2061 2074 7970 6963 616c 202e 424f 4f0d   a typical .BOO.
000006e0: 6669 6c65 206c 6f6f 6b20 6c69 6b65 3a0d  file look like:.
000006f0: 0d20 2043 4b45 524d 4954 2e45 5845 0d20  .  CKERMIT.EXE. 
00000700: 2043 4559 5030 4964 3035 4030 507e 336f   CEYP0Id05@0P~3o
00000710: 6f6d 6f32 5930 3146 5765 5038 4030 3037  omo2Y01FWeP8@007
00000720: 5030 3030 3034 3048 4234 3030 3160 577e  P000040HB4001`W~
00000730: 3238 624c 3030 355c 577e 324a 4250 3030  28bL005\W~2JBP00
00000740: 3732 3256 305a 4850 5950 3a0d 2020 5c38  722V0ZHPYP:.  \8
00000750: 3a48 325d 5232 5630 5b60 5059 503a 3638  :H2]R2V0[`PYP:68
00000760: 3e48 3253 3233 5630 5948 5069 503a 5867  >H2S23V0YHPiP:Xg
00000770: 3830 303b 5164 7e32 5557 4430 3036 5967  800;Qd~2UWD006Yg
00000780: 7e32 4f67 6c30 3039 5d6f 7e32 4c38 3030  ~2Ogl009]o~2L800
00000790: 303b 3230 7e7e 7e7e 0d20 207e 7e7e 7e7e  0;20~~~~.  ~~~~~
000007a0: 7e7e 3a52 3248 3030 3854 563f 5037 3631  ~~:R2H008TV?P761
000007b0: 5434 3130 3c48 3640 5034 306a 346c 3652  T410<H6@P40j4l6R
000007c0: 5248 3030 3833 6c31 3740 5050 3f60 314d  RH0083l17@PP?`1M
000007d0: 403f 5953 5032 306f 3045 6530 6e55 4430  @?YSP20o0Ee0nUD0
000007e0: 6833 6c0d 2020 3157 4433 6a4f 4033 5d30  h3l.  1WD3jO@3]0
000007f0: 566a 5730 333d 384c 3f58 3460 4e30 6f30  VjW03=8L?X4`N0o0
00000800: 3130 6f31 5d65 375b 4032 5c50 4f3d 384c  10o1]e7[@2\PO=8L
00000810: 4836 3040 3030 5261 6a3e 3034 5e39 3758  H60@00Raj>04^97X
00000820: 6830 0d0d 4578 6563 7574 6162 6c65 2043  h0..Executable C
00000830: 2d4b 6572 6d69 7420 7072 6f67 7261 6d73  -Kermit programs
00000840: 2061 7265 2064 6973 7472 6962 7574 6564   are distributed
00000850: 2069 6e20 424f 4f20 666f 726d 6174 2066   in BOO format f
00000860: 6f72 2074 6865 2043 6f6d 6d6f 646f 7265  or the Commodore
00000870: 0d41 6d69 6761 2c20 4f53 2f32 2c20 616e  .Amiga, OS/2, an
00000880: 6420 706f 7373 6962 6c79 206f 7468 6572  d possibly other
00000890: 732e 2020 424f 4f20 666f 726d 6174 2069  s.  BOO format i
000008a0: 7320 616c 736f 2075 7365 6420 666f 7220  s also used for 
000008b0: 4d53 2d44 4f53 204b 6572 6d69 742c 0d4d  MS-DOS Kermit,.M
000008c0: 6963 726f 736f 6674 2057 696e 646f 7773  icrosoft Windows
000008d0: 204b 6572 6d69 742c 2061 6e64 206f 7468   Kermit, and oth
000008e0: 6572 204b 6572 6d69 7420 7665 7273 696f  er Kermit versio
000008f0: 6e73 2074 6861 7420 6172 6520 6e6f 7420  ns that are not 
00000900: 7061 7274 206f 660d 432d 4b65 726d 6974  part of.C-Kermit
00000910: 2e0d 0d54 6865 2070 726f 6772 616d 2066  ...The program f
00000920: 6f72 2063 6f6e 7665 7274 696e 6720 6120  or converting a 
00000930: 6269 6e61 7279 2065 7865 6375 7461 626c  binary executabl
00000940: 6520 286f 7220 616e 7920 6f74 6865 7229  e (or any other)
00000950: 2066 696c 6520 696e 746f 0d42 4f4f 2066   file into.BOO f
00000960: 6f72 6d61 7420 6973 3a0d 0d20 2043 4b42  ormat is:..  CKB
00000970: 4d4b 422e 430d 0d43 6f6d 7069 6c65 2074  MKB.C..Compile t
00000980: 6869 7320 7072 6f67 7261 6d20 7573 696e  his program usin
00000990: 6720 796f 7572 2043 2063 6f6d 7069 6c65  g your C compile
000009a0: 7220 616e 6420 7468 656e 2072 756e 2069  r and then run i
000009b0: 742c 2067 6976 696e 6720 7468 6520 6e61  t, giving the na
000009c0: 6d65 0d6f 6620 7468 6520 696e 7075 7420  me.of the input 
000009d0: 2862 696e 6172 7929 2061 6e64 206f 7574  (binary) and out
000009e0: 7075 7420 2862 6f6f 2920 6669 6c65 7320  put (boo) files 
000009f0: 6173 2061 2063 6f6d 6d61 6e64 2d6c 696e  as a command-lin
00000a00: 6572 0d65 7861 6d70 6c65 3a0d 0d20 2063  er.example:..  c
00000a10: 6b62 6d6b 6220 636b 6572 6d69 742e 6578  kbmkb ckermit.ex
00000a20: 6520 636b 6572 6d69 742e 626f 6f0d 0d54  e ckermit.boo..T
00000a30: 6865 2070 726f 6772 616d 2066 6f72 2063  he program for c
00000a40: 6f6e 7665 7274 696e 6720 6120 424f 4f20  onverting a BOO 
00000a50: 6669 6c65 2062 6163 6b20 696e 746f 2069  file back into i
00000a60: 7473 206f 7269 6769 6e61 6c20 666f 726d  ts original form
00000a70: 2028 7375 6368 2061 730d 6269 6e61 7279   (such as.binary
00000a80: 2065 7865 6375 7461 626c 6529 2069 733a   executable) is:
00000a90: 0d0d 2020 434b 4255 4e42 2e43 0d0d 436f  ..  CKBUNB.C..Co
00000aa0: 6d70 696c 6520 7468 6973 2070 726f 6772  mpile this progr
00000ab0: 616d 2075 7369 6e67 2079 6f75 7220 4320  am using your C 
00000ac0: 636f 6d70 696c 6572 2061 6e64 2074 6865  compiler and the
00000ad0: 6e20 7275 6e20 6974 2c20 6769 7669 6e67  n run it, giving
00000ae0: 2074 6865 206e 616d 650d 6f66 2074 6865   the name.of the
00000af0: 2069 6e70 7574 2028 626f 6f29 2066 696c   input (boo) fil
00000b00: 6520 6173 2061 2063 6f6d 6d61 6e64 2d6c  e as a command-l
00000b10: 696e 6520 6172 6775 6d65 6e74 2c20 666f  ine argument, fo
00000b20: 7220 6578 616d 706c 653a 0d0d 2020 636b  r example:..  ck
00000b30: 6275 6e62 2063 6b65 726d 6974 2e62 6f6f  bunb ckermit.boo
00000b40: 0d0d 5468 6973 2077 696c 6c20 7265 2d63  ..This will re-c
00000b50: 7265 6174 6520 7468 6520 6669 6c65 2077  reate the file w
00000b60: 6974 6820 6974 7320 6f72 6967 696e 616c  ith its original
00000b70: 206e 616d 6520 2861 7320 7374 6f72 6564   name (as stored
00000b80: 2069 6e20 7468 6520 6669 7273 740d 6c69   in the first.li
00000b90: 6e65 206f 6620 7468 6520 424f 4f20 6669  ne of the BOO fi
00000ba0: 6c65 292e 0d0d 5468 6573 6520 7477 6f20  le)...These two 
00000bb0: 7072 6f67 7261 6d73 2061 7265 2069 6465  programs are ide
00000bc0: 6e74 6963 616c 2074 6f20 7468 6520 4d53  ntical to the MS
00000bd0: 424d 4b42 2061 6e64 204d 5342 5043 5420  BMKB and MSBPCT 
00000be0: 7072 6f67 7261 6d73 2074 6861 7420 6172  programs that ar
00000bf0: 650d 6469 7374 7269 6275 7465 6420 7769  e.distributed wi
00000c00: 7474 2e20 2042 4f4f 2d66 696c 6520 6d61  tt.  BOO-file ma
00000c10: 6b65 7273 2061 6e64 2064 6563 6f64 6572  kers and decoder
00000c20: 7320 6172 6520 616c 736f 0d61 7661 696c  s are also.avail
00000c30: 6162 6c65 2069 6e20 6f74 6865 7220 6c61  able in other la
00000c40: 6e67 7561 6765 7320 2846 6f72 7472 616e  nguages (Fortran
00000c50: 2c20 5061 7363 616c 2c20 4173 7365 6d62  , Pascal, Assemb
00000c60: 6c65 722c 2065 7463 2920 696e 2074 6865  ler, etc) in the
00000c70: 204d 532d 444f 530d 4b65 726d 6974 2066   MS-DOS.Kermit f
00000c80: 696c 6520 636f 6c6c 6563 7469 6f6e 2075  ile collection u
00000c90: 6e64 6572 2074 6865 204d 5342 2070 7265  nder the MSB pre
00000ca0: 6669 782e 0d5e 430d 0a                   fix..^C..
```

**Reply by dbalsom** [reply]
Here are the differences (besides changed line endings):
![image](https://github.com/dbalsom/martypc/assets/7229541/e7d869f0-899f-4d9d-b72c-748d1ce9d457)

Rather than a byte here and there, it looks like we lose blocks of characters at a time. 

This is probably more on the passthrough logic than on Marty's serial port. That stuff was a quick hack and could probably use another look and possible refactor. It's tricky to interface between a device that runs in real time and one that runs in a virtual time slice.  One cheat I could do is give the 8250 an input FIFO only in passthrough mode, that might help with the overruns.

**Reply by cr1901** [reply]
It very well could be a `com0com` bug/limitation. But then I can't reliably use `com0com` for data transfer, and we're back at my original problem :P.

>One cheat I could do is give the 8250 an input FIFO only in passthrough mode, that might help with the overruns.

If proper emulation is a concern (since 8250 doesn't have a FIFO), could the FIFO be placed in between the virtual serial port passthrough (COM22) and the start of the 8250 emulation (COM2)?

**Reply by dbalsom** [reply]
> If proper emulation is a concern (since 8250 doesn't have a FIFO), could the FIFO be placed in between the virtual serial port passthrough (COM22) and the start of the 8250 emulation (COM2)?

I'd probably rewrite the serial passthrough to use crossbeam channels, and we'd just recv() a new byte once one was read out. Then no need to explicitly emulate a FIFO buffer as the bytes will just hang out in the channel. I need to move everything to channels anyway with the eventual goal of getting the core running in a separate thread from the gui.

**Reply by cr1901** [reply]
Another option for file transfer is of course "create a floppy image from a directory tree on host", and do the reverse to "extract to host". But I'd still be interested in the COM port xfer since it's a broadly applicable data xfer technique (and serial ports are fun :D).

**Reply by dbalsom** [reply]
> Another option for file transfer is of course "create a floppy image from a directory tree on host", and do the reverse to "extract to host". 

![image](https://github.com/user-attachments/assets/5c892c45-dd80-42b9-9897-d2434cdcb63a)

i aim to please

**Reply by cr1901** [reply]
Well, this solves my original problem :). I guess using virtual serial ports will just have to remain "not supported" on Windows, since it seems that `com0com` is dropping chars before Marty ever sees them.
